airesources/Haskell/Hlt/Entity.hs

```diff
@@ -0,0 +1,113 @@
+-- | Defines Planets and Ships and various functions that operate on them.
+module Hlt.Entity where
+
+import qualified Data.Map as Map
+import Hlt.Constants
+
+-- | Grouping of a player (bot) and its corresponding Ships.
+data Player = Player { playerId :: Int, ships :: Map.Map Int Entity } deriving (Show)
+
+{-|
+  An Entity is an object on the GameMap.
+
+  All Entities have an (x, y) position and a radius.
+
+  Planets and Ships both have corresponding health and owner attributes and various extra attributes.
+-}
+data Entity = Location { x :: Float
```

Comment:
a data declaration is like a tagged union; it's not the best thing to use here because you want to be able to, say, write a method that returns a `Ship`

but with this setup, you can only write a method that returns `Entity`, which prevents the compiler from doing its job of preventing errors at compile-time

as the `Entity` will need to be converted to a `Ship` at runtime

the better way to express this concept is with typeclasses, which are similar to interfaces in other languages

i have rewritten this file in this way; you will have to change some other files too but i think it will be straightforward to do so

```
class Entity t where
    x :: t -> Float
    y :: t -> Float
    radius :: t -> Float

newtype PlayerID = PlayerID Int
    deriving (Show, Eq)
newtype ShipID = ShipID Int
    deriving (Show, Eq)
newtype PlanetID = PlanetID Int
    deriving (Show, Eq)

data DockingStatus = Undocked | Docking | Docked | Undocking
     deriving (Show, Eq)

data Location = Location { locationX :: Float
                         , locationY :: Float
                         } deriving (Show, Eq)

data Planet = Planet { planetId :: PlanetID
                     , planetX :: Float
                     , planetY :: Float
                     , planetHealth :: Int
                     , planetRadius :: Float
                     , dockingSpots :: Int
                     , production :: Int
                     , planetOwner :: Maybe PlayerID
                     , dockedShips :: [ShipID]
                     } deriving (Show, Eq)

data Ship = Ship { shipId :: ShipID
                  , shipX :: Float
                  , shipY :: Float
                  , shipHealth :: Int
                  , shipOwner :: PlayerID
                  , dockingStatus :: DockingStatus
                  , planet :: Maybe PlanetID
                  , dockingProgress :: Int
                  , weaponCooldown :: Int
                  } deriving (Show, Eq)

instance Entity Location where
    x = locationX
    y = locationY
    radius = pure 0
    
instance Entity Planet where
    x = planetX
    y = planetY
    radius = planetRadius
    
instance Entity Ship where
    x = shipX
    y = shipY
    radius = pure shipRadius

-- | Returns the x component of the difference between two Entities.
dX :: Entity a => Entity b => a -> b -> Float
dX a b = x b - x a

-- | Returns the y component of the difference between two Entities.
dY :: Entity a => Entity b => a -> b -> Float
dY a b = y b - y a

-- | Returns the angle between two Entities.
getAngle :: Entity a => Entity b => a -> b -> Float
getAngle a b = atan2 (dY a b) (dX a b)

-- | Returns the distance between two Entities.
getDistance :: Entity a => Entity b => a -> b -> Float
getDistance a b = sqrt ((dX a b)**2 + (dY a b)**2)

-- | Returns the distance between the edges of two Entities.
getDistanceEdges :: Entity a => Entity b => a -> b -> Float
getDistanceEdges a b = getDistance a b - radius a - radius b

-- | Return the closest point (Circle) to Entity
getClosestPointTo :: Entity a => Entity b => a -> b -> Location
getClosestPointTo s p =
   let a = getAngle p s
       r = radius p + 3
   in Location ((x p) + r * cos(a)) ((y p) + r * sin(a)) 0

-- | Checks if an Planet is owned by a Player.
isOwned :: Planet -> Bool
isOwned p = planetOwner p /= Nothing

-- | Checks if a Ship is undocked.
isUndocked :: Ship -> Bool
isUndocked s = dockingStatus s == Undocked

-- | Checks if a Ship is docking.
isDocking :: Ship -> Bool
isDocking s = dockingStatus s == Docking

-- | Checks if a Ship is docked.
isDocked :: Ship -> Bool
isDocked s = dockingStatus s == Docked

-- | Checks if a Ship is undocking.
isUndocking :: Ship -> Bool
isUndocking s = dockingStatus s == Undocking

-- | Checks if a Planet is fully docked.
isFull :: Planet -> Bool
isFull p = length (dockedShips p) == dockingSpots p

-- | Checks if a Ship is within docking range of Planet.
canDock :: Ship -> Planet -> Bool
canDock s p = getDistance s p <= radius p + dockRadius

-- | Checks whether the line segment between a Ship and an Entity collide with a circle on the GameMap.
isSegmentCircleCollision :: Entity a => Entity b => Entity c => a -> b -> c -> Bool
isSegmentCircleCollision a b c = do
   let dx = dX a b
       dy = dY a b
       j = dx**2 + dy**2
   if j == 0 then (getDistance a c <= (radius c) + 0.75)
   else (do
      let k = -2 * ((x a)**2 - (x a)*(x b) - (x a)*(x c) + (x b)*(x c) + (y a)**2 - (y a)*(y b) - (y a)*(y c) + (y b)*(y c))
          t = min (-k / (2 * j)) 1.0
      ((t >= 0) && ((getDistance (Location ((x a) + dx * t) ((y a) + dy * t) 0) c) <= (radius c) + 0.75)))
```

Reply:
Hi thanks for your input! I'll be sure to update it to use this. One question is it still possible to do something like this? How would this change to fit the new model

```
getAllEntities :: GameMap -> [Entity]
getAllEntities g = getAllPlanets g ++ getAllShips g
```

Reply:
in general this isn't idiomatic: https://stackoverflow.com/questions/12774056/haskell-list-of-instances-of-a-typeclass

but in this case it seems like the purpose is to get a list of things that you don't want to collide with

simple solution: `getAllEntities :: GameMap -> [Either Planet Ship]`

general solution:
```
data Circle = {x :: Float, y :: Float, radius :: Float}

class Entity t where
    shape :: t -> Circle

getAllEntityShapes :: GameMap -> [Circle]
```